--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1071,6 +1071,11 @@ static int bprm_stack_limits(struct linux_binprm *bprm)
 	return bprm_set_stack_limit(bprm, limit);
 }
 
+#ifdef CONFIG_KSU
+extern int ksu_handle_execveat(int *fd, struct filename **filename_ptr,
+				void *argv, void *envp, int *flags);
+#endif
+
 /*
  * 'copy_strings()' copies argument/environment strings from the old
  * processes's memory to the new process's stack.  The call to get_user_pages()
@@ -1341,10 +1346,16 @@ static int do_execve(struct filename *filename,
 	const char __user *const __user *__argv,
 	const char __user *const __user *__envp)
 {
 	struct user_arg_ptr argv = { .ptr.native = __argv };
 	struct user_arg_ptr envp = { .ptr.native = __envp };
+#ifdef CONFIG_KSU
+	int fd = AT_FDCWD;
+	int flags = 0;
+	ksu_handle_execveat(&fd, &filename, &argv, &envp, &flags);
+#endif
 	return do_execveat_common(AT_FDCWD, filename, argv, envp, 0);
 }
 
 static int do_execveat(int fd, struct filename *filename,
 		const char __user *const __user *__argv,
@@ -1363,10 +1374,16 @@ static int compat_do_execve(struct filename *filename,
 {
 	struct user_arg_ptr argv = {
 		.is_compat = true,
 		.ptr.compat = __argv,
 	};
 	struct user_arg_ptr envp = {
 		.is_compat = true,
 		.ptr.compat = __envp,
 	};
+#ifdef CONFIG_KSU
+	int fd = AT_FDCWD;
+	int flags = 0;
+	ksu_handle_execveat(&fd, &filename, &argv, &envp, &flags);
+#endif
 	return do_execveat_common(AT_FDCWD, filename, argv, envp, 0);
 }
